src/main/java/net/ripe/rpki/rsyncit/rrdp/RrdpFetcher.java

```diff
@@ -240,15 +245,13 @@ private ProcessPublishElementResult processPublishElements(Element doc, Optional
                     // https://www.w3.org/TR/2004/PER-xmlschema-2-20040318/datatypes.html#base64Binary
                     var decoded = decoder.decode(content.trim());
 
-                    var hash = Sha256.asBytes(decoded);
-
-                    // Try to get some creation timestamp from the object itself. If it's impossible to parse
-                    // the object, use the default based on the last-modified header of the snapshot.
                     //
                     // Cache the timestamp per hash do avoid re-parsing every object in the snapshot every time.
                     //
-                    final Instant modificationTime = state.cacheTimestamps(Sha256.asString(hash), now,
-                        () -> incorporateHashInTimestamp(getTimestampForObject(objectUri, decoded, defaultTimestamp), hash));
+                    // We can not use hashes in sub-second precision because rsync may start syncing those by default.
```

Comment:
The default_time + hash % N is only used if we fail to extract timestamp from the signing time. So hash here is to be  able to distinguish (meaning, tell the difference based just on the file modification time) between different objects with the same name and different content. 

If rsync starts to use nanoseconds it will be comparing these hash-based modification times with much higher precision. Why is it a problem?

Reply:
There is this trick that validators apply for failover. If the validator writes the files to filesystem with a time that is the same as on the server, the file will not be transferred when fallback happens.

The main one implementing this is rpki-client.

If we use nanoseconds _and_ rsync starts transfering this precision, this trick no longer works and we need to transfer all data.

Reply:
I see
But does that mean we should better mimic rpki-client behaviour precisely?
Because both (hash -or-> last-modified -or-> 00:00 of the current hour) and (last-modified -or-> 00:00 of the current hour) are probably different from what rpki-client does. And I'm not even sure what it does with objects it can't get signing time from.

Reply:
My read of the code before my change is that we incorporate the hash into the timestamp **also** when the signing time can be extracted. That makes this change applicable.

Reply:
Yeah, you are right, it was a bug actually (that never popped up).
The hash was supposed to be applied to the defaultTimestamp not to the object time 🙄 
Anyway, that is settled now